src/plugins/kibana_react/public/app_links/click_handler.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import { ApplicationStart } from 'src/core/public';
+import { getClosestLink, hasActiveModifierKey } from './utils';
+
+interface CreateCrossAppClickHandlerOptions {
+  container: HTMLElement;
+  currentAppId: string;
+  navigateToApp: ApplicationStart['navigateToApp'];
+  parseAppUrl: ApplicationStart['parseAppUrl'];
+}
+
+export const createCrossAppClickHandler = ({
+  container,
+  currentAppId,
+  navigateToApp,
+  parseAppUrl,
+}: CreateCrossAppClickHandlerOptions): React.MouseEventHandler<HTMLElement> => {
+  return (e) => {
+    if (container == null) {
+      return;
+    }
+    // see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/12239
+    const target = e.target as HTMLElement;
+
+    const link = getClosestLink(target, container);
+    if (!link) {
+      return;
+    }
+
+    if (
+      link.href && // ignore links with empty hrefs
+      (link.target === '' || link.target === 'self') && // ignore links having a target
+      e.button === 0 && // ignore everything but left clicks
+      !e.defaultPrevented && // ignore default prevented events
+      !hasActiveModifierKey(e) // ignore clicks with modifier keys
+    ) {
+      const appInfo = parseAppUrl(link.href);
+      if (appInfo && appInfo.app !== currentAppId) {
```

Comment:
Remind me again why we can't call this for links within the same app?

Reply:
It'd be nice if this solution worked for all links

Reply:
It's an opinionated decision, based on the fact that most (all) out apps are currently using `react-router` and the `Link` component. Handling links to the active link could/would result in a click on the link be handled both by the `Link` click handler and ours, resulting on performing the `history.push` twice. (Also the name of our component is `redirect**Cross**AppLinks`

But I'm unsure about that TBH, maybe we should just decide than any app links should be handled and document that using `Link` inside `RedirectCrossAppLinks` should be avoided? wdyt @joshdover @Dosant ?

Reply:
Looking at https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/modules/Link.js#L49, the `Link` onClick does prevent default, so it would just no trigger our own handler and works properly. I guess I can remove the currentApp check then. 

Should the component be renamed to `RedirectAppLinks` then?

Reply:
> based on the fact that most (all) out apps are currently using `react-router` and the `Link` component

I'm not sure this is true? In my rudimentary search (`git grep "import.*Link.*from 'react-router*" -- {src,x-pack}/plugins`) I only found that beats_management and uptime plugins were using this component. I think the main reason devs are not using this component is that you can't really use it with EuiButton or EuiLink.

> We are probably fine to remove the currentApp check then. Which also means we can use the component for the navlinks.
> 
> Should the component be renamed to `RedirectAppLinks` then?

If it works, than a rename makes sense to me 👍